Headcrab rework

Robustin4 opened the pull request:
## What Does This PR Do
Реворк хедкрабов. Теперь они что-то вроде мидраунд антагонистов, гоняют по вентам, их логова теперь спавнятся по всей станции, можно брать под контроль, макаки теперь тоже могут быть зомби (баг был у меня только или всегда он был). ((в коде лежат еще два хедкраба, но для них нет спрайтов, да и я их сделал))

## Why It's Good For The Game
Новая угроза станции.

Robustin3 committed:
хедкрабсреворк1

аыв

Robustin3 committed:
еще одно дерьмо кашалота

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -52,14 +56,33 @@
 		new spawn_type(T)
+		successSpawn = TRUE
+
+	var/c = rand(2,4)
```
Что за невероятное "c"

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -52,14 +56,33 @@
+	var/c = rand(2,4)
+
+	while(turfs.len > 0 && c > 0)
+		var/turf/simulated/floor/T = pick(availableareas)
```
Что за невероятное "T"

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -52,14 +56,33 @@
+
+	if(successSpawn)
+		var/list/spawners = list()
+		for(var/obj/structure/spawner/headcrab/he in world)
```
he-who?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -11,24 +11,141 @@
+			to_chat(src, "To jump, try to use ranged attack.")
+		else
+			to_chat(src, "<span class='notice'>You are a zombie!</span>")
+			to_chat(src, "Maybe you will try to kill humans for bodys for your little brothers?")
```
bodys -> bodies

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -11,24 +11,141 @@
+		return
+
+	if(!QDELETED(candidate) || !QDELETED(candidate.mob))
+		var/datum/mind/M = create_headcrab_mind(candidate.ckey)
```
Невероятное M, почему не user, к примеру

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -11,24 +11,141 @@
+			to_chat(src, "<span class='notice'>You are a zombie!</span>")
+			to_chat(src, "Maybe you will try to kill humans for bodys for your little brothers?")
+
+/proc/create_headcrab_mind(key)
```
А точно нужен глобальный прок для этого?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -11,24 +11,141 @@
+	set name = "Hide"
+	set desc = "Become invisible to the common eye."
+
+	if(is_zombie)
```
А почему бы не убирать этот верб у зомби?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -11,24 +11,141 @@
+
 
 /mob/living/simple_animal/hostile/headcrab/Life(seconds, times_fired)
+
```
Лишние пробелы можно было бы и убрать

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -39,9 +156,73 @@
 						health += 10
 					qdel(K)
 					break
+	if(!ai)
```
AI, в плане, станционный? Если ИИ игровой, то почему бы не дать просто проверку на клиента?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -39,9 +156,73 @@
+
+	if(prob(1)) //it was a proc with a name its_time_to_kill_yourself
+
+		var/turf/T2 = get_turf(src) //как тебе такое, волмат? двухбуквенная переменная!
```
Очень не очень

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -39,9 +156,73 @@
+					break
+
+	if(travelling_in_vent)
+		if(istype(loc, /turf))
```
isturf()

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -39,9 +156,73 @@
+
+	if(prob(15) && vent_cooldown <= 0)
+		if(!is_zombie || is_zombie && host_species == "Monkey" || "Farwa" || "Neara" || "Stok" || "Wolpin")
+			for(var/obj/machinery/atmospherics/unary/vent_pump/v in view(4,src))
```
Внизу ты же сделал temp_vent, почему тут просто невероятная v?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -39,9 +156,73 @@
+				return
+			var/obj/machinery/atmospherics/unary/vent_pump/exit_vent = pick(vents)
+			if(prob(75))
+				visible_message("<B>[src] scrambles into the ventilation ducts!</B>", \
```
Точно visible, а не audible?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -63,17 +260,27 @@
 	icon = H.icon
 	speak = list('sound/creatures/zombie_idle1.ogg','sound/creatures/zombie_idle2.ogg','sound/creatures/zombie_idle3.ogg')
 	speak_chance = 50
+	obj_damage += 40
```
+=? Точно следует добавлять это условие? Хедкраб помогает ручками ломать объекты?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -83,7 +290,11 @@
+	visible_message("<span class='warning'>[H.name]'s body suddenly rises!")
+
+	to_chat(src, "<span class='notice'>You are a zombie now!</span>")
+	to_chat(src, "No ventcrawling now (if you are not monkey and etc). But more health.")
```
Можно было добавить с условием с проверкой на ventrcrawling

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
+
+	speed = 2.45
+
+	obj_damage += 40
```
тоже самое, что выше

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
 	. = ..()
-	if(iscarbon(target) && target.reagents)
+
+	if(. && neurotoxin_per_jump > 0 && iscarbon(target) && target.reagents && !is_zombie)
```
Так neurotoxin_per_jump используется только тут, верно? Зачем ее пихать в hostile/headcrab, а не только в hostile/headcrab/poison?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
+			C.reagents.add_reagent("headcrabneurotoxin", neurotoxin_per_jump)
+
+	if(src.is_zombie && isturf(src.loc) && src.poison_headcrabs != 0)
+		for(var/mob/living/carbon/human/H in oview(src, 4)) //Only for carbons on poison headcrab jumping distance
```
Как много внизу однобуквенных переменных. Кто кого убивает? Кто жертва, кто нападающий?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
+
+/mob/living/simple_animal/hostile/headcrab/reviver //no sprites, but coded //x2
+	name = "reviver headcrab"
+	desc = "A strange parasitic creature that would like to connect with your brain stem through your upper body. It's looks like, that it can shock you."
```
It's looks like -> It looks like

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
+
+	playsound(src.loc, pick('sound/effects/sparks1.ogg', 'sound/effects/sparks2.ogg', 'sound/effects/sparks3.ogg'), 20, 1)
+
+	..()
```
А это точно нужно, если в начале уже есть .=..()?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -181,13 +401,152 @@
+
+	R.revive_cooldown += 30
+
+	return ..()
```
нет новой строки в конце файла, и зачем return ..()?

Robustin4 commented:
Ладно, спасибо, вправлю код, когда будет время.

PhantornRU commented:
Не называй так коммиты. Можешь хоть сотню повторов того чего делаешь вписывать. Но хуевые названия коммитов - признак плохого тона.
![image](https://user-images.githubusercontent.com/41479614/187812467-ed58685b-05ee-4541-a9b5-b07b2c36e3d8.png)

Robustin3 committed:
запросы и еще че то

ага, запросы ларентоун и эти... небольшие штучки сделал

PhantornRU commented:
@Robustin4 то что ты сделал по запросам - внизу нажимаешь:
![image](https://user-images.githubusercontent.com/41479614/187956846-21152ccb-15db-4b4b-b381-45dbe88a82b8.png)

Robustin3 committed:
правкин

Правкин сделал правочки

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -446,11 +447,11 @@
-	if(src.is_zombie && isturf(src.loc) && src.poison_headcrabs != 0)
-		if(src.poison_headcrabs != 0 && src.is_zombie)
+	if(src.is_zombie && isturf(src.loc) && src.poison_headcrabs != 0) // в оригинале несколько хедркабов было на спине у ядовитого, и еще... он ими кидался. у нас же он их внезапно рожает и умеет кидаться
+		if(src.poison_headcrabs != 0 && src.is_zombie) //игрок кидаться может всегда, а ИИ лишь с шансом.
```
Проверка is_zombie выше уже есть

Robustin3 committed:
ЛарентоунЗапросус

Robustin3 committed:
стоп нахера я удваивал условие...

видимо это было из-за усталости, когда я болел

Robustin3 committed:
поливрока (невнимательность в коде)

ага.

PhantornRU reviewed (commented):
Ну йобобаны.  Ну непонятно ж из таких коммитов что было сделано.
Рофлы это конечно хорошо, но нужно вкратце хотя бы писать чем ты был занят в этом ПРе, даже если это просто правки в коде.
![image](https://user-images.githubusercontent.com/41479614/188346528-3b7c36f0-0401-41b4-958f-0a1e3b9b46a2.png)

Robustin3 committed:
эволюция? (не проверено фактически, лишь теоретически)

Robustin3 committed:
логическая правка

писал же, теоритически работает, фактически не знаю, все лишь на теории смотрю, вот заметил логическую ошибку

Robustin3 committed:
Правка правки

Действительно.

Robustin3 committed:
правка однобуквенности переменной

немного пропустил, когда запрос ларентоун делал

Robustin3 committed:
Отхил и Правки

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -12,7 +12,7 @@
 	ranged = 1
 	ventcrawler = 2
-	ranged_message = "leaps"
+	ranged_message = "бросается"
```
Перевод запрещен

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -149,9 +150,9 @@
-		health = maxHealth
 		projectiletype = /obj/item/projectile/toxinvomit
 		projectilesound = 'sound/weapons/pierce.ogg'
+		ranged_message = "блюет"
```
Перевод запрещен

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -242,15 +243,38 @@
 				if(faction_check_mob(L) && !attack_same)
 					return
-	visible_message("<span class='danger'><b>[src]</b> [ranged_message] at [A]!</span>")
+	visible_message("<span class='danger'><b>[src]</b> [ranged_message] на [A]!</span>")
```
Перевод запрещен, особенно без склонений!

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -637,4 +662,4 @@
 
-	headcrab.revive_cooldown += 30
\ No newline at end of file
+	headcrab.revive_cooldown += 30
```
end of line есть?

Robustin3 committed:
Точка-бесючка и приз робустам если выживут долго

Robustin3 committed:
АнтиПеревод и еще eof

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -156,7 +156,7 @@
 		projectiletype = /obj/item/projectile/toxinvomit
 		projectilesound = 'sound/weapons/pierce.ogg'
-		ranged_message = "блюет"
+		ranged_message = "puking"
```
pukes?

Robustin4 replied:
![image](https://user-images.githubusercontent.com/111193615/188937772-7c598d73-728b-4a67-b3c9-1d72f23cbd2c.png)

Robustin3 committed:
Фикс Грамматики

Robustin3 committed:
Правки && Ивент && Гонарх(без спрайтов)

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+		if(turf_clear(thefloor))
+			turfs += thefloor
+
+	while(turfs.len > 0 && how_many_capsules > 0)
```
length(turfs) использовать лучше

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+		sleep(2)
+		how_many_capsules--
+		var/turf/simulated/floor/where_capsule = pick(turfs)
+		//explosion(where_capsule, 0.2, 0.2,0.2) // взрыв все ломает. пожалуй откажемся
```
А вы зачем ставите деци-значения, если это радиус в тайлах?

Robustin4 replied:
не разбираюсь в взрывах.

larentoun replied:
Ну и ладненько. Просто на будущее - цифорки это в тайлах. Радиус гиба, сильный, слабый.

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+		var/sound/global_boom = sound('sound/weapons/mortar_long_whistle.ogg')
+
+		for(var/P in GLOB.player_list)
+			var/mob/M = P
```
Можно было и нормальное название переменной

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+		var/obj/structure/crabmissile/capsule = new /obj/structure/crabmissile(where_capsule)
+		headcrabs_release(null, capsule, rand(0,1)/*, TRUE*/)
+
+/datum/event/crabmissiles/proc/headcrabs_release(var/mob/living/simple_animal/hostile/headcrab/headcrab_type, var/obj/structure/crabmissile/capsule, var/randomized_headcrabs/*, var/admin_log*/)
```
Зачем оставлять комментированный код?

Robustin4 replied:
да так, на будущее

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+		headcrab_type = pick(/mob/living/simple_animal/hostile/headcrab, /mob/living/simple_animal/hostile/headcrab/fast, /mob/living/simple_animal/hostile/headcrab/poison)
+	var/capsule_position = get_turf(capsule)
+
+	//выдает ошибку про координаты. отказался.
```
Опять, ненужный комментированный код

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+
+	while(headcrabs_in_capsule > 0)
+		headcrabs_in_capsule--
+		sleep(40)
```
А не хотите использовать addtimer. Они более оптимизированы

Robustin4 replied:
я не совсем понимаю принцип его работы

larentoun replied:
Вызывает прок через время

Robustin4 replied:
я не понял как же его вызвать необходимое количество раз...

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+	for(var/turf/simulated/floor/nest_probably_position in range(capsule, 4))
+		nest_probably_position -= capsule_position
+		var/turf/simulated/floor/nest_position = pick(nest_probably_position)
+		new/obj/structure/spawner/headcrab/(nest_position)
```
А тут точно нормально описан new?

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
+	notify_ghosts("Появились хедкрабы", source = thecapsule, action = NOTIFY_ATTACK, flashwindow = FALSE)
+
+/datum/event/crabmissiles/announce()
+	//GLOB.event_announcement.Announce("Внимание, [station_name()]. Было зафиксировано приближение к вам [how_many_capsules == 1 ? "[how_many_capsules]" : " "] [how_many_capsules == 1 ? "капсулы" : "капсул" ], содержащие паразитов Хедкрабы. Готовьтесь к обороне и ремонту. Слава НаноТрейзен!", "Сообщение от ОСН 'Фаланга'") //йуху, ивент, прошедший на веге. ссылается на какую-то ошибку про двоеточие.
```
Комментированный код

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -68,22 +68,189 @@
 #undef HEADCRAB_POISON
 #undef HEADCRAB_MEGAMIX
-#undef HEADCRAB_SPAWNER
+#undef HEADCRAB_SPAWNER
```
Убрал EOL? Или как?

Robustin4 replied:
почему-то ругается

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -36,6 +37,7 @@
 	var/gonome = FALSE
 	var/can_be_gonomed = TRUE
 	var/gonome_time = 468
+	var/is_gonarch = 1 // кажется уже какой-то спагетти код, прямо как писал американец ниже.
```
TRUE/FALSE может лучше?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -666,4 +677,90 @@
+	obj_damage = 264
+	armour_penetration = 15
+	environment_smash = 3
+	attacktext = "пронзает"
```
Перевод запрещен

Robustin4 replied:
странно, attacktext хедкрабов переведен, почему бы не продолжить традицию?

larentoun replied:
Отмени перевод

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -666,4 +677,90 @@
+		var/obj/item/stack/sheet/metal/materials = new /obj/item/stack/sheet/metal(dismantle_location)
+		qdel(src)
+		materials.amount = 26
+		TOOL_DISMANTLE_SUCCESS_MESSAGE
```
EOL есть?

Robustin4 replied:
есть

Robustin3 committed:
Вновь запрос. Вновь выполнение.

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -199,7 +194,7 @@
 		nest_probably_position -= capsule_position
 		var/turf/simulated/floor/nest_position = pick(nest_probably_position)
-		new/obj/structure/spawner/headcrab/(nest_position)
+		new /obj/structure/spawner/headcrab/(nest_position)
```
Компилится то проект? По-моему не должно быть `/` в headcrab/(nest_position)

Robustin4 replied:
компилируется, а / забыл убрать

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -699,11 +699,11 @@
 	armour_penetration = 15
 	environment_smash = 3
-	attacktext = "пронзает"
+	attacktext = "pierces" //по просьбе ларентоун отменил перевод. вообще не логично, это я создал этого моба, и я по-русски написал его аттак текст. где перевод? я вообще мог подшутить и оставить наследование attacktext от хедкраба. однако грызть гонарх не может...
```
Я писала про то, чтобы откатывать перевод (еще неделю назад, наверное, когда там ревью были), и там вроде откатывал?

Robustin4 replied:
Перевод запрещен. Ладно.

Robustin3 committed:
ночные правки

Robustin3 committed:
Хайпоп зависимость и правочки

Robustin3 committed:
language color

larentoun commented:
О, тебе линтеры включили. Почитай, там две ошибки

Robustin3 committed:
ребаланс и возможый багфикс

Robustin3 committed:
проверка клиента по-другому

Robustin3 committed:
чета руагается, че-то чиница

Robustin4 commented:
И как решить этот линтер? Я же с базой данных никак не контактировал...

Robustin3 committed:
logging capsule landing

Robustin3 committed:
правочки малявочки

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -143,7 +143,7 @@
 		to_chat(src, "Now you can shoot toxic vomit, healed for 25 health and have additional 50 health of maximum.")
 		ranged = 1
-		ranged_cooldown_time = 125
+		ranged_cooldown_time = 666
```
Минута? Ничего не перепутал?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -636,6 +638,7 @@
 	if(prob(95))
 		var/mob/living/carbon/C = target
 		C.adjustFireLoss(12)
+		C.adjustStaminaLoss(5) //stuncrab
```
5 урона по стамине до 100 нужных?

Robustin3 committed:
Ошибочка и Нерфочка

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -646,7 +646,7 @@
 		var/mob/living/carbon/C = target
 		C.adjustFireLoss(12)
-		C.adjustStaminaLoss(5) //stuncrab
+		C.adjustStaminaLoss(2) //stuncrab
```
Я про то, что нужно 100 стамина урона, чтобы довести до стаминакрита... А приписка stuncrab говорит о том, что он должен быть эффективным станящим, нет? 50 ударов для стана, ух

Robustin4 replied:
Ну если я поставлю большое значение - дисбаланс
Что нибудь придумаю.

larentoun replied:
У нас в среднем 3-4 удара на стан (3 без брони, 4 с броней), плюс-минус 35 урона по стамине.

larentoun replied:
Ну это уже решать тебе/через педальсостав/по желанию левой пятки, но я всё еще прослежу, чтобы не было слишком сильно/слабо

Robustin3 committed:
Logic and Stuncrab

Robustin3 committed:
LOGIC

Robustin3 committed:
Logicson

Robustin3 committed:
Logic & Role Pref || Role Ban && misc

Robustin3 committed:
Баланс что ли...

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -1,3 +1,6 @@
+
+var/global/hctypes = "/mob/living/simple_animal/hostile/headcrab, /mob/living/simple_animal/hostile/headcrab/fast, /mob/living/simple_animal/hostile/headcrab/poison"
```
Глобальные переменные определяются через GLOB.

Robustin4 replied:
а что не так через var/global/ ?

larentoun replied:
> а что не так через var/global/ ?

var/global/ это не глобальная переменная. Советую почитать contributing.md

Robustin3 committed:
Logic. Again.

Robustin3 committed:
forgot one thing

Robustin3 committed:
GLOB, not var/global and little things

Robustin3 committed:
LittleThings

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -251,6 +251,19 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+
+	var/turf/death_loc = get_turf(src)
+
+	var/mob/living/simple_animal/hostile/headcrab/headcrabb = new src(death_loc) //уподобление оригиналу, этот вид хедкраба должен быть убит вне тела.
```
headcrab`b`?

Robustin3 committed:
Непонятки какие-то. Правки.

Robustin3 committed:
Патч

larentoun commented on code/datums/ruins/lavaland.dm:
```diff
@@ -226,3 +226,12 @@
+	suffix = "lavaland_ruin_headcrab_lab.dmm"
+	allow_duplicates = FALSE
+	cost = 10 // nanomed, guns, security autolathe, atmos, plasma and fat loot.
+	//always_place = TRUE // tests
```
Надеюсь, потом уберешь

Robustin4 replied:
я тестировал ранее, потом закомментировал, думаю убирать ничего не надо

larentoun replied:
Лишние комментарии в коде - плохо. Кому надо, тот сам вернет

Robustin4 replied:
ладна

larentoun commented on code/game/objects/items/devices/scanners.dm:
```diff
@@ -388,6 +388,12 @@ REAGENT SCANNER
 	if(H.has_brain_worms())
 		. += "<span class='warning'>Обнаружено отклонение в мозговой активности."
 		. += "&emsp;Рекомендуется подробное сканирование.</span>"
+	if(H.reagents.has_reagent("headcrab_neurotoxin"))
```
А нам точно такое нужно? Это же просто токсин, нет?

Robustin4 replied:
Идем к оригиналу: этот нейротоксин понижал здоровье героя до 10 или 1, затем вводилось противоядие, думаю медсканер не так уж и далеко уйдет от костюма Гордона, причем токсин довольно опасен

larentoun replied:
Может оформишь его как вирус, как вариант? Как раз поддержка есть всего этого.

Robustin4 replied:
ммммм хороша идея

Robustin4 replied:
Вот только вирусом... Он будет постоянным, а у нас нейротоксин, чье действие можно легко остановить, и лишь когда он в организме.

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -93,6 +93,39 @@
+
+	var/list/candidates = SSghost_spawns.poll_candidates("Хотите стать одиноким хедкрабом?", ROLE_HEADCRAB, TRUE, poll_time = 100, source = headcrab_type)
+
+	if(!candidates.len)
```
```suggestion
	if(!length(candidates))
```

Robustin4 replied:
Не вижу разницы, но сделаю

larentoun commented on code/modules/events/headcrabs.dm:
```diff
@@ -93,6 +93,39 @@
 
 /*
 
+==================================
```
Можно убрать?

Robustin4 replied:
Как будет время за компьютером

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -22,27 +22,36 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 	var/is_gonarch = FALSE // кажется уже какой-то спагетти код, прямо как писал американец ниже.
+	var/is_monkey_type = FALSE
+
+/mob/living/simple_animal/hostile/headcrab/Move(atom/newloc, dir, step_x, step_y)
```
А зачем пустой прок?

Robustin4 replied:
Судя по твоему фиксу хождения серого паука ужаса по диагонали через стекла - этот прок нужен

larentoun replied:
А, ну, если у хостайлов/симплов прок Мува имеет другие переменные, тогда могло быть и ладно, НО!
![image](https://user-images.githubusercontent.com/31931237/196005899-54082109-6b22-43be-ae93-6c6ef17dbc1e.png)
Я глянул - у хостайлов ровно такой же прок. Вывод: по идее лишнее и ничего не меняет

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -113,6 +118,18 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	else
+		infest = TRUE
+
+	to_chat(src, "<span class=notice'>You are [infest ? "toggled" : "disabled" ] infestation of body.</span>")
```
```suggestion
	to_chat(src, "<span class=notice'>You have [infest ? "toggled" : "disabled" ] infestation of body.</span>")
```

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -130,10 +147,15 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 		hiding = FALSE
 
+/mob/living/simple_animal/hostile/headcrab/adjustHealth(damage)
+	..()
```
Уверен, что не нужно возвращать?

Robustin4 replied:
Ну без возвращения нормально работает

larentoun replied:
Глянул в коде, ты обязан вернуть родителя, так как он используется как минимум в калькуляции урона

Robustin4 replied:
Ладно, выпущу его с родительского собрания

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -286,10 +317,16 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 	src.verbs -= /mob/living/simple_animal/hostile/headcrab/verb/hide_headcrab
 	src.verbs -= /mob/living/simple_animal/hostile/headcrab/verb/build_a_nest
+	src.verbs -= /mob/living/simple_animal/hostile/headcrab/verb/infest_disable
+	src.verbs += /mob/living/verb/pulled
```
А ты его где-то меняешь? Ответь тут

Robustin4 replied:
Кого? Где?

larentoun replied:
`src.verbs += /mob/living/verb/pulled`

Robustin4 replied:
/mob/living/simple_animal/hostile/headcrab/New()

 ...
 
 	if(!is_zombie)
		verbs -= /mob/living/verb/pulled

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -301,30 +338,18 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 	H.update_hair()*/
 	host_species = H.dna.species.name
-	if(host_species == "Monkey" || "Farwa" || "Neara" || "Stok" || "Wolpin")
+	if(ismonkeybasic(H) || isfarwa(H) || iswolpin(H) || isneara(H) || isstok(H))
```
Не нужно сравнение с `host_species`? Если не нужно, зачем тебе переменная `host_species`?

Robustin4 replied:
Для иконок. С host_species что то не работало

larentoun replied:
Тогда убери `host_species`

Robustin4 replied:
Проверка иконок зомби все равно требует host_species. Оставляем.

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -405,7 +430,7 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 
 	if(istype(firer, /mob/living/simple_animal/hostile/headcrab/poison))
 		victim.adjustToxLoss(rand(2,8))
```
Что за `victim`?

Robustin4 replied:
Жертва

larentoun replied:
Бл- Кхм... Зачем она тут, почему две разные переменные означают одно и тоже?

Robustin4 replied:
А хрен его знает, поправлю, когда будет возможность.

Robustin4 replied:
А, понял в чем причина твоего негодования... Ну... Э.. Мне с target я хоть уверен, что нейротоксин будет в организме карбона, victim меня раньше подводил

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -443,8 +469,13 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	move_to_delay = 5.6
 
-	maxHealth -= 20
+	var/newname = "fast [name]"
```
А переменная точно нужна?

Robustin4 replied:
Да.

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -575,21 +620,23 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 	jumpspeed = 0.5
 	speak_emote = list("slowly hisses")
-	damage_coeff = list(BRUTE = 0.88, BURN = 0.88)
+	damage_coeff = list(BRUTE = 0.88, BURN = 0.88, TOX = 1, CLONE = 1, STAMINA = 0, OXY = 1)
```
Уточнять `= 1` обязательно?

Robustin4 replied:
Из-за этого краб становится неуязвимым к другим источникам урона

larentoun replied:
А, ну да, всё верно.
Тебе стоило сделать
```suggestion
	damage_coeff[BRUTE] = 0.88
	damage_coeff[BURN] = 0.88
```

Robustin4 replied:
code\modules\mob\living\simple_animal\hostile\headcrab.dm:667:error: damage_coeff: left-hand side must be an object variable

larentoun commented on code/modules/reagents/chemistry/reagents/toxins.dm:
```diff
@@ -27,19 +27,82 @@
+		var/turf/T = get_turf(M)
+		T.add_vomit_floor()
+
+	if(prob(chance))
```
У тебя дважды может эмоция проиграться. Сделай список и из него `pick()`

larentoun commented on code/modules/reagents/chemistry/reagents/toxins.dm:
```diff
@@ -27,19 +27,82 @@
+		var/mob/living/carbon/human/M2 = M
+		M2.set_heartattack(TRUE)
+
+	if(prob(chance))
```
Тут вообще ужас всё, что ниже. Подсмотри, как у зависимостей шанс работает и проки на различные эффекты

Robustin4 replied:
И так сойдёт

larentoun replied:
> И так сойдёт

Не сойдет, просто тяжести зафиксируй

Robustin4 replied:
что как

larentoun replied:
Я же сказал - посмотри пример с зависимостей

Robustin4 replied:
Ладно

Robustin3 committed:
патч2

Robustin3 committed:
патч3

larentoun commented on code/modules/reagents/chemistry/reagents/toxins.dm:
```diff
@@ -57,11 +57,11 @@
-
-	if(prob(chance))
-		M.emote("groan")
+	if(prob(chance) + 3.2)
```
Что за невероятные 3.2? У тебя 100% будут играться эмоции

Robustin4 replied:
а вот такие, 6.8 + 3.2

larentoun replied:
Ты знаешь, как работает условие if?

Robustin4 replied:
Ну дааааааа

Robustin3 committed:
патч4

Robustin3 committed:
багфикс и жрачка

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 
+	return ..()
+
+/mob/living/simple_animal/hostile/headcrab/attack_hand(mob/living/carbon/human/M as mob)
```
as mob не нужен

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	return ..()
+
+/mob/living/simple_animal/hostile/headcrab/attack_hand(mob/living/carbon/human/M as mob)
+	..()
```
Зачем ты дважды вызываешь родителя?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	return ..()
+
+/obj/item/reagent_containers/food/snacks/headcrab
+
```
лишний пробел

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	var/armored = FALSE
+
+/obj/machinery/cooker/foodgrill/putIn(obj/item/In, mob/chef)
+	..()
```
зачем ты дважды вызываешь родителя

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+
+	..()
+
+/obj/item/reagent_containers/food/snacks/headcrab/attack(mob/M, mob/user, def_zone)
```
Что за `mob/M`?

Robustin4 replied:
Судя по логике, M это тот, кого кормят. Больше идей не имею. Остальное к писарю легаси кода

larentoun replied:
> Судя по логике, M это тот, кого кормят. Больше идей не имею. Остальное к писарю легаси кода

Поменял бы тогда. Погоди, ты сам не понимаешь, что делаешь?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	..()
+
+/obj/item/reagent_containers/food/snacks/headcrab/attack(mob/M, mob/user, def_zone)
+	..()
```
Уверен, что сначала нужно вызвать родителя?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	if(istype(src, /obj/item/reagent_containers/food/snacks/headcrab) && src.armored)
+		if(do_after(src, 40, target = user, progress=TRUE))
+			src.armored = FALSE
+			to_chat(user, "<span class='notice'>His armor was breaked! Time to eat!</span>")
```
```suggestion
			to_chat(user, "<span class='notice'>His armor is now broken! Time to eat!</span>")
```

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+		return
+
+/obj/item/reagent_containers/food/snacks/headcrab/attack_self(mob/user)
+	..()
```
Точно нужно вызывать родителя сначала?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	qdel(src)
+
+	if(prob(6))
+		to_chat(M, "<span class='notice'>Strange... Headcrab tastes not like crab...</span>")
```
```suggestion
		to_chat(M, "<span class='notice'>Strange... Headcrabs don't taste like crabs...</span>")
```

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+		to_chat(M, "<span class='notice'>Strange... Headcrab tastes not like crab...</span>")
+
+/mob/living/simple_animal/hostile/headcrab/proc/scoop_up_headcrab(var/mob/living/carbon/grabber)
+
```
Лишний пробел?

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+
+/mob/living/simple_animal/hostile/headcrab/proc/scoop_up_headcrab(var/mob/living/carbon/grabber)
+
+	var/obj/item/reagent_containers/food/snacks/headcrab/H = new /obj/item/reagent_containers/food/snacks/headcrab(loc)
```
Ты уже прописал путь, просто `= new(loc)`

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	H.desc = desc
+
+	if(istype(H, /mob/living/simple_animal/hostile/headcrab/poison) || istype(H, /mob/living/simple_animal/hostile/headcrab/reviver))
+		desc += " Eating and/or cooking this - a very bad idea."
```
```suggestion
		desc += "It is a very bad idea to eat or cook this."
```

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+/mob/living/carbon/attackby(obj/item/I, mob/user, params)
+	..()
+
+	if(istype(src, /mob/living/simple_animal/hostile/headcrab) && src.stat == DEAD && istype(I, /obj/item/storage/bag/trash/))
```
```suggestion
	if(istype(src, /mob/living/simple_animal/hostile/headcrab) && src.stat == DEAD && istype(I, /obj/item/storage/bag/trash))
```

larentoun replied:
Ты себе разве не делал дефайна `isheadcrab`?

Robustin4 replied:
Нет.

Robustin4 replied:
если я уберу "/" то не будет проверка также и на другие виды мешков для мусора - блюспейсовый и киборгский

larentoun replied:
> если я уберу "/" то не будет проверка также и на другие виды мешков для мусора - блюспейсовый и киборгский

В плане? Не совсем понял тебя

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	return H
+
+/mob/living/carbon/attackby(obj/item/I, mob/user, params)
+	..()
```
Уверен, что в начале нужно это?

Robustin4 replied:
Да.

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -315,10 +315,109 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
+	to_chat(grabber, "<span class='notice'>You scoop up \the [src].")
+	return H
+
+/mob/living/carbon/attackby(obj/item/I, mob/user, params)
```
Объясни логику. Если карбона ударили чем-то, И если этот карбон хедкраб...?????

Robustin4 replied:
Карбон ударил хедкраба мешком для мусора - хедкраб в этом мешке для мусора

larentoun replied:
> Карбон ударил хедкраба мешком для мусора - хедкраб в этом мешке для мусора

Так тут логика "когда ударили карбона айтемом I"

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -667,6 +766,10 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 /mob/living/simple_animal/hostile/headcrab/armored/attack_hand(mob/living/carbon/human/M)
+
+	if(src.stat == DEAD)
+		return
```
Уверен, что не нужно возвращать функцию родителя?

Robustin3 committed:
Жрачка: Логикал Аддитионал

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -367,14 +367,17 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 			src.armored = FALSE
 			to_chat(user, "<span class='notice'>His armor was breaked! Time to eat!</span>")
+			var/number = 0
+			number = rand(1,6)
```
Я либо слепой, либо не вижу, где оно используется

Robustin4 replied:
playsound(user.loc, "sound/effects/bone_break[number].ogg", 25, TRUE, ignore_walls = FALSE, is_global = null)

Robustin3 committed:
йоу ларри просил ларри получил

Robustin3 committed:
Логикалли2

Robustin3 committed:
return ..()

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -320,15 +321,13 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
-/mob/living/simple_animal/hostile/headcrab/attack_hand(mob/living/carbon/human/M as mob)
-	..()
+/mob/living/simple_animal/hostile/headcrab/attack_hand(mob/living/carbon/human/M)
 
```
Лишние пробелы то убирай

larentoun commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -338,7 +337,6 @@ GLOBAL_LIST_INIT(hctypes, list(/mob/living/simple_animal/hostile/headcrab, /mob/
 
 /obj/machinery/cooker/foodgrill/putIn(obj/item/In, mob/chef)
-	..()
 
```
Лишние пробелы

Robustin3 committed:
приказ ларри

Robustin3 committed:
переносченко
